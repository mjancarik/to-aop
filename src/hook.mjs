import { AOP_FILTER_FUNCTION } from './symbol.mjs';

export function createHook(name, regular, callback) {
  function hook(meta) {
    return hookFor(meta, regular, callback);
  }

  hook[AOP_FILTER_FUNCTION] = function ({ property, target, prototype }) {
    return (
      (typeof regular === 'string' && property.includes(regular)) ||
      (regular instanceof RegExp && regular.test(property)) ||
      (typeof regular === 'function' &&
        regular({ property, target, name, prototype }))
    );
  };

  return {
    [name]: hook,
  };
}

export function hookFor(meta, regular, callback) {
  if (typeof regular === 'string') {
    if (meta.property.includes(regular)) {
      return callback(meta);
    }

    return null;
  }
  if (regular instanceof RegExp) {
    if (regular.test(meta.property)) {
      return callback(meta);
    }

    return null;
  }

  if (typeof regular === 'function') {
    if (regular(meta)) {
      return callback(meta);
    }

    return null;
  }

  if (Array.isArray(regular)) {
    return regular.map(({ rule, action }) => {
      return hookFor(meta, rule, action || callback);
    });
  }

  throw new TypeError(
    `Invalid rule type ${typeof regular}. Method accept string, regexp, function and array.`,
  );
}

export function hasToRegisterHook(hooks) {
  return (pattern, props) => {
    return hooks.reduce((result, hook) => {
      return (
        result ||
        (pattern[hook] &&
          typeof pattern[hook][AOP_FILTER_FUNCTION] === 'function' &&
          pattern[hook][AOP_FILTER_FUNCTION](props))
      );
    }, false);
  };
}

export const hookName = Object.freeze({
  beforeMethod: 'beforeMethod',
  afterMethod: 'afterMethod',
  aroundMethod: 'aroundMethod',
  beforeGetter: 'beforeGetter',
  afterGetter: 'afterGetter',
  aroundGetter: 'aroundGetter',
  beforeSetter: 'beforeSetter',
  afterSetter: 'afterSetter',
  aroundSetter: 'aroundSetter',
});
