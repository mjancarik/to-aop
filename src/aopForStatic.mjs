import createGetTrap from './trap/createGetTrap.mjs';
import createSetTrap from './trap/createSetTrap.mjs';
import overOwnProperty from './overOwnProperty.mjs';
import { AOP_STATIC_ALLOW, AOP_HOOKS } from './symbol.mjs';
import { hookName, hasToRegisterHook } from './hook.mjs';

const hasToRegisterGetterSetterHook = hasToRegisterHook([
  hookName.beforeGetter,
  hookName.afterGetter,
  hookName.aroundGetter,
  hookName.beforeSetter,
  hookName.afterSetter,
  hookName.aroundSetter,
]);

export default function aopForStatic(target, pattern) {
  let original = {};
  let originalTarget = target;

  if (!Object.prototype.hasOwnProperty.call(originalTarget, AOP_STATIC_ALLOW)) {
    Reflect.defineProperty(originalTarget, AOP_STATIC_ALLOW, {
      value: false,
      enumerable: false,
      writable: true,
    });
  } else {
    originalTarget[AOP_STATIC_ALLOW] = false;
  }

  while (target && target !== Function.prototype) {
    // TODO improve
    Object.entries(Object.getOwnPropertyDescriptors(target)).forEach(
      ([property, descriptor]) => {
        if (
          typeof descriptor.get === 'function' ||
          typeof descriptor.set === 'function'
        ) {
          if (Object.prototype.hasOwnProperty.call(original, property)) {
            return;
          }

          Reflect.defineProperty(original, property, descriptor);
          if (
            !hasToRegisterGetterSetterHook(pattern, {
              target,
              property,
              object: original,
            })
          ) {
            return;
          }

          Reflect.defineProperty(
            target,
            property,
            Object.assign({}, descriptor, {
              get: (...rest) => {
                if (originalTarget[AOP_STATIC_ALLOW] === true) {
                  let aopHooks = descriptor.get[AOP_HOOKS];
                  if (aopHooks) {
                    const { object } = aopHooks[aopHooks.length - 1];

                    aopHooks.push({
                      target,
                      object,
                      property,
                      pattern,
                    });

                    return typeof descriptor.get === 'function'
                      ? descriptor.get(...rest)
                      : undefined;
                  }

                  return createGetTrap({
                    target,
                    object: original,
                    property,
                    pattern,
                  })(...rest);
                } else {
                  return typeof descriptor.get === 'function'
                    ? descriptor.get(...rest)
                    : undefined;
                }
              },
              set: (payload) => {
                if (originalTarget[AOP_STATIC_ALLOW] === true) {
                  return createSetTrap({
                    target,
                    object: original,
                    property,
                    pattern,
                  })(payload);
                }
              },
            }),
          );
        }
      },
    );

    overOwnProperty({ target, pattern, original, object: target });
    target = Reflect.getPrototypeOf(target);
  }

  originalTarget[AOP_STATIC_ALLOW] = true;
}
