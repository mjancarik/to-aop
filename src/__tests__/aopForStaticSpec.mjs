import { expect, jest, describe, beforeEach, it } from '@jest/globals';

import aopForStatic from '../aopForStatic.mjs';
import { hookName } from '../hook.mjs';
import createClasses from './createClasses.mjs';
import createPattern from './createPattern.mjs';

describe('aopForStatic method', () => {
  let pattern = null;
  let pattern2 = null;

  let beforeMethod = null;
  let afterMethod = null;
  let beforeGetter = null;
  let afterGetter = null;
  let beforeSetter = null;
  let afterSetter = null;

  beforeEach(() => {
    beforeMethod = jest.fn();
    afterMethod = jest.fn();
    beforeGetter = jest.fn();
    afterGetter = jest.fn();
    beforeSetter = jest.fn();
    afterSetter = jest.fn();

    pattern = createPattern(undefined, {
      [hookName.beforeMethod]: beforeMethod,
      [hookName.afterMethod]: afterMethod,
      [hookName.beforeGetter]: beforeGetter,
      [hookName.afterGetter]: afterGetter,
      [hookName.beforeSetter]: beforeSetter,
      [hookName.afterSetter]: afterSetter,
    });
    pattern2 = createPattern(undefined, {
      [hookName.beforeMethod]: beforeMethod,
      [hookName.afterMethod]: afterMethod,
      [hookName.beforeGetter]: beforeGetter,
      [hookName.afterGetter]: afterGetter,
      [hookName.beforeSetter]: beforeSetter,
      [hookName.afterSetter]: afterSetter,
    });
  });

  it('should call pattern.beforeMethod and pattern.afterMethod with multiple aspect', () => {
    let { C } = createClasses();

    aopForStatic(C, pattern);
    aopForStatic(C, pattern2);
    const staticResult1 = C.staticMethod();
    const staticResult2 = C.staticMethod2();

    expect(beforeMethod.mock.calls.length).toEqual(4);
    expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
    expect(afterMethod.mock.calls.length).toEqual(4);
    expect(afterMethod.mock.calls[0]).toMatchSnapshot();

    expect(staticResult1).toMatchInlineSnapshot(`"static method"`);
    expect(staticResult2).toMatchInlineSnapshot(`"static method 2"`);
  });

  it('should create meta information from pattern.beforeMethod and pass to pattern.afterMethod', () => {
    const value = 'value';
    beforeMethod = ({ meta }) => (meta.information = 'value');
    afterMethod = ({ meta }) => expect(meta.information).toEqual(value);

    let { C } = createClasses();

    aopForStatic(C, pattern);

    C.staticMethod();
  });

  it('should call pattern.beforeMethod and pattern.afterMethod for own static method with multiple aspect', () => {
    let { C, A } = createClasses();

    aopForStatic(A, pattern);
    aopForStatic(C, pattern);
    const staticResult2 = C.staticMethod2();

    expect(beforeMethod.mock.calls.length).toEqual(1);
    expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
    expect(afterMethod.mock.calls.length).toEqual(1);
    expect(afterMethod.mock.calls[0]).toMatchSnapshot();

    expect(staticResult2).toMatchInlineSnapshot(`"static method 2"`);
  });

  it('should call pattern.beforeMethod and pattern.afterMethod for inhereted static method with multiple aspect', () => {
    let { C, A } = createClasses();

    aopForStatic(A, pattern);
    aopForStatic(C, pattern);

    const staticResult = C.staticMethod();

    expect(beforeMethod.mock.calls.length).toEqual(2);
    expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
    expect(afterMethod.mock.calls.length).toEqual(2);
    expect(afterMethod.mock.calls[0]).toMatchSnapshot();

    expect(staticResult).toMatchInlineSnapshot(`"static method"`);
  });

  it('should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter', () => {
    let { C } = createClasses();

    aopForStatic(C, pattern);

    const staticResult1 = C.staticGetter;

    expect(staticResult1).toMatchInlineSnapshot(`"static getter"`);

    expect(beforeGetter.mock.calls.length).toEqual(1);
    expect(beforeGetter.mock.calls[0]).toMatchSnapshot();
    expect(afterGetter.mock.calls.length).toEqual(1);
    expect(afterGetter.mock.calls[0]).toMatchSnapshot();
  });

  it('should create meta information from pattern.beforeGetter and pass to pattern.afterGetter', () => {
    const value = 'value';
    beforeGetter = ({ meta }) => (meta.information = 'value');
    afterGetter = ({ meta }) => expect(meta.information).toEqual(value);
    pattern = createPattern(undefined, {
      [hookName.beforeGetter]: beforeGetter,
      [hookName.afterGetter]: afterGetter,
    });

    let { C } = createClasses();

    aopForStatic(C, pattern);

    C.staticGetter;
  });

  it('should not call pattern.beforeGetter and pattern.afterGetter for not hooked property', () => {
    let hookFilterMethod = ({ property }) => property === 'staticGetter2';

    let pattern = createPattern(
      {
        [hookName.beforeGetter]: hookFilterMethod,
        [hookName.afterGetter]: hookFilterMethod,
        [hookName.beforeSetter]: hookFilterMethod,
        [hookName.afterSetter]: hookFilterMethod,
      },
      {
        [hookName.beforeGetter]: beforeGetter,
        [hookName.afterGetter]: afterGetter,
        [hookName.beforeSetter]: beforeSetter,
        [hookName.afterSetter]: afterSetter,
      },
    );
    let { C } = createClasses();

    aopForStatic(C, pattern);

    const staticResult1 = C.staticGetter;

    expect(staticResult1).toMatchInlineSnapshot(`"static getter"`);

    expect(beforeGetter.mock.calls.length).toEqual(0);
    expect(beforeGetter.mock.calls[0]).toMatchSnapshot();
    expect(afterGetter.mock.calls.length).toEqual(0);
    expect(afterGetter.mock.calls[0]).toMatchSnapshot();
  });

  it('should not throw error after call aopForStatic', () => {
    let { D } = createClasses();

    expect(() => {
      aopForStatic(D, pattern);
    }).not.toThrow();
  });

  it('should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter with multiple aspect', () => {
    let { A, C } = createClasses();

    aopForStatic(A, pattern);
    aopForStatic(C, pattern);

    const staticResult1 = C.staticGetter;

    expect(staticResult1).toMatchInlineSnapshot(`"static getter"`);

    expect(beforeGetter.mock.calls.length).toEqual(2);
    expect(beforeGetter.mock.calls[0]).toMatchSnapshot();
    expect(afterGetter.mock.calls.length).toEqual(2);
    expect(afterGetter.mock.calls[0]).toMatchSnapshot();
  });

  it('should call pattern.beforeSetter and pattern.afterSetter for inhereted static setter', () => {
    let { C } = createClasses();

    aopForStatic(C, pattern);

    C.staticSetter = 'static setter';

    expect(C.staticSetter).toMatchInlineSnapshot(`"static setter"`);

    expect(beforeSetter.mock.calls.length).toEqual(1);
    expect(beforeSetter.mock.calls[0]).toMatchSnapshot();
    expect(afterSetter.mock.calls.length).toEqual(1);
    expect(afterSetter.mock.calls[0]).toMatchSnapshot();
  });

  it('should create meta information from pattern.beforeSetter and pass to pattern.afterSetter', () => {
    const value = 'value';
    beforeSetter = ({ meta }) => (meta.information = 'value');
    afterSetter = ({ meta }) => expect(meta.information).toEqual(value);
    pattern = createPattern(undefined, {
      [hookName.beforeSetter]: beforeSetter,
      [hookName.afterSetter]: afterSetter,
    });

    let { C } = createClasses();

    aopForStatic(C, pattern);

    C.staticSetter = 'static setter';
  });

  it('should not call pattern.afterMethod for static getter which return constructable function', () => {
    let { C, A } = createClasses();

    C.A = A;

    aopForStatic(C, pattern);

    const a = Reflect.construct(C.A, []);

    expect(a instanceof A).toBeTruthy();

    expect(afterMethod.mock.calls.length).toEqual(0);
    expect(afterMethod.mock.calls[0]).toMatchSnapshot();
  });
});
