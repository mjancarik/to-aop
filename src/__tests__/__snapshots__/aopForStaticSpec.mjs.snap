// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter 2`] = `
[
  {
    "context": undefined,
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter 3`] = `
[
  {
    "context": undefined,
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static getter",
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter with multiple aspect 2`] = `
[
  {
    "context": undefined,
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter with multiple aspect 3`] = `
[
  {
    "context": undefined,
    "meta": {},
    "object": {
      "staticMethod": [Function],
    },
    "payload": "static getter",
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for inhereted static method with multiple aspect 1`] = `
[
  {
    "args": [],
    "context": [Function],
    "meta": {},
    "object": {
      "staticMethod": [Function],
    },
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for inhereted static method with multiple aspect 2`] = `
[
  {
    "args": [],
    "context": [Function],
    "meta": {},
    "object": {
      "staticMethod": [Function],
    },
    "payload": "static method",
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for own static method with multiple aspect 1`] = `
[
  {
    "args": [],
    "context": [Function],
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticMethod2",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for own static method with multiple aspect 2`] = `
[
  {
    "args": [],
    "context": [Function],
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static method 2",
    "property": "staticMethod2",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod with multiple aspect 1`] = `
[
  {
    "args": [],
    "context": [Function],
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod with multiple aspect 2`] = `
[
  {
    "args": [],
    "context": [Function],
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static method",
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeSetter and pattern.afterSetter for inhereted static setter 2`] = `
[
  {
    "context": undefined,
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static setter",
    "property": "staticSetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeSetter and pattern.afterSetter for inhereted static setter 3`] = `
[
  {
    "context": undefined,
    "meta": {},
    "object": {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static setter",
    "property": "staticSetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should not call pattern.afterMethod for static getter which return constructable function 1`] = `undefined`;

exports[`aopForStatic method should not call pattern.beforeGetter and pattern.afterGetter for not hooked property 2`] = `undefined`;

exports[`aopForStatic method should not call pattern.beforeGetter and pattern.afterGetter for not hooked property 3`] = `undefined`;
